### PR TITLE
status: Registering Prometheus Collectors on Status-plugin reconfigure

### PR DESCRIPTION
**Pull request by johanfylling**
if `Prometheus` was enabled on active config change; and unregistering collectors if disabled.

Fixes: #5918

### CODE REVIEW

File: plugins/status/plugin.go [hunk 1]
```diff
@@ -258,6 +261,24 @@ func (p *Plugin) register(r prom.Registerer, cs ...prom.Collector) {
...
+	if p.manager.PrometheusRegister() != nil {
+		p.register(p.manager.PrometheusRegister(), allCollectors...)
+	}
+}
+
+func (p *Plugin) unregister(r prom.Registerer, cs ...prom.Collector) {
+	for _, c := range cs {
+		r.Unregister(c)
+	}
+}
```
Comment (philipaconrad) [inline plugins/status/plugin.go:274]: [comment]: This is a clever trick for getting some API symmetry between `register`/`unregister`. :smile:

File: plugins/status/metrics.go [hunk 2]
```diff
@@ -60,6 +60,18 @@ var (
 		Help:    "Histogram for the bundle loading duration by stage.",
 		Buckets: prometheus.ExponentialBuckets(1000, 2, 20),
 	}, []string{"name", "stage"})
+
+	allCollectors = []prometheus.Collector{
```
Comment (philipaconrad) [inline plugins/status/metrics.go:64]: [suggestion]: Would a comment/note here along the lines of "if you add new metrics, make sure to add them here, or they won't survive status plugin reconfigure events" be worthwhile? I feel like there's a small tripping hazard lurking here, but it's not a blocker either way.

**Review by philipaconrad — DISMISSED** [review]
This looks good to me! I spent some time staring at the refactored reconfig code, since there's a potential concurrency hazard there around channel waits/sends.

I do like that the problem of "knowing when we're done reconfiguring the plugin" is solved in a fairly straightforward way, using that `done` signaling channel. :slightly_smiling_face: 

Only a few comments/suggestions below:

File: plugins/status/plugin.go [hunk 3]
```diff
@@ -378,8 +401,9 @@ func (p *Plugin) loop() {
 					p.logger.Info("Status update sent successfully in response to discovery update.")
 				}
 			}
-		case newConfig := <-p.reconfig:
-			p.reconfigure(newConfig)
+		case update := <-p.reconfig:
+			p.reconfigure(update.config)
+			update.done <- struct{}{}
```
[the remark this answers is not in the document]
Comment (johanfylling) [reply plugins/status/plugin.go:406]: The reason is that this loop is happening in a go routine, and we want to block in `Reconfigure()` outside that routine. We could safely be guarding this with a mutex here instead, I think, but then we'd need to come up with another way of testing this without spooking the race-condition tests. We have [precedence](https://github.com/open-policy-agent/opa/blob/77475dc2ab1db6349925d1c020a3b134bc3f37e7/plugins/logs/plugin.go#LL770C11-L770C11) of doing this in the log plugin, so I don't think there is any harm other than a slight delay at the caller of `Reconfigure()`.

[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply plugins/status/plugin.go:406]: Thanks for the clarification.

**Review by ashutosh-narkar — APPROVED** [review]
LGTM

**Pull request merged**